setup/src/setup-haskell.ts

```diff
@@ -21,34 +29,106 @@ export default async function run(
   try {
     core.info('Preparing to setup a Haskell environment');
     const os = process.platform as OS;
+    const home = process.env.HOME;
     const opts = getOpts(getDefaults(os), os, inputs);
 
+    // Cache parameters
+    let cachePaths: string[] = [];
+    let hashFile = '';
+
     for (const [t, {resolved}] of Object.entries(opts).filter(o => o[1].enable))
       await core.group(`Installing ${t} version ${resolved}`, async () =>
         installTool(t as Tool, resolved, os)
       );
 
-    if (opts.stack.setup)
+    if (opts.stack.setup) {
       await core.group('Pre-installing GHC with stack', async () =>
         exec('stack', ['setup', opts.ghc.resolved])
       );
+    }
 
-    if (opts.cabal.enable)
+    if (opts.stack.enable) {
+      const stackRoot = os === 'win32' ? 'C:\\sr' : `${home}/.stack`;
+      core.setOutput('stack-root', stackRoot);
+      if (os === 'win32') core.exportVariable('STACK_ROOT', 'C:\\sr');
+
+      if (opts.cache) {
+        const matches = await glob
+          .create('**/stack.*.lock')
+          .then(async g => g.glob());
+        if (!matches[0]) await exec('stack', ['ls', 'dependencies']);
+        cachePaths = [stackRoot];
+        hashFile = matches[0] || 'stack.yaml.lock';
+      }
+    }
+
+    if (opts.cabal.enable) {
       await core.group('Setting up cabal', async () => {
         await exec('cabal', ['user-config', 'update'], {silent: true});
         const configFile = await cabalConfig();
 
-        if (process.platform === 'win32') {
+        if (os === 'win32')
           fs.appendFileSync(configFile, 'store-dir: C:\\sr\n');
-          core.setOutput('cabal-store', 'C:\\sr');
-        } else {
-          core.setOutput('cabal-store', `${process.env.HOME}/.cabal/store`);
-        }
 
         await exec('cabal user-config update');
         if (!opts.stack.enable) await exec('cabal update');
       });
+      const cabalStore = os === 'win32' ? 'C:\\sr' : `${home}/.cabal/store`;
+      core.setOutput('cabal-store', cabalStore);
+
+      if (opts.cache) {
+        const matches = await glob
+          .create('**/cabal.*.freeze')
```

Comment:
Odd question: How does this work if

1. A haskell project uses cabal and the project is both an exe and a library.
2. Said project has a cabal.project.freeze file as per exe recommendations, but names it something different by default so that building with `cabal build` can be tested to work if the project is used as a library.

What if the project does not want to cache the `dist-newstyle` directory and only wants to cache the cabal store? It's a reasonable want for some projects. Examples include: for a while, macOS was suffering dylib issues when caching `dist-newstyle` and so I had to turn off caching dist-newstyle until that was resolved; it can take up a *lot* of extra space with not much extra benefit, especially if you're using submodules for dependencies instead of "vendoring" them with a `cabal.project` file (you pay the penalty of having the submodules bloat dist-newstyle as well as not having mtime work well so you frequently have to rebuild all of them anyway even if it is cached; but if you use `cabal.project` to vendor them, they get built and stored in the cabal.store, making caching dist-newstyle more effective).

[dist-newstyle](https://github.com/haskell/cabal/issues/5731) is also not guaranteed to remain the same name forever. So there's that...